chain/network/src/peer_manager/network_state/tier1.rs

```diff
@@ -77,4 +179,147 @@ impl super::NetworkState {
+        for d in accounts_data.data.values() {
+            proxies_by_account.entry(&d.account_key).or_default().extend(d.proxies.iter());
+            for p in &d.proxies {
+                accounts_by_proxy.entry(&p.peer_id).or_default().push(&d.account_key);
+            }
+        }
+
+        // Browse the connections from newest to oldest.
+        let tier1 = self.tier1.load();
+        let mut ready: Vec<_> = tier1.ready.values().collect();
+        ready.sort_unstable_by_key(|c| c.established_time);
+        ready.reverse();
+
+        // Select the oldest TIER1 connection for each account.
+        let mut safe = HashMap::<&PublicKey, &PeerId>::new();
+        if validator_cfg.is_some() {
+            // TIER1 nodes can also connect to TIER1 proxies.
+            for conn in &ready {
+                let peer_id = &conn.peer_info.id;
+                for key in accounts_by_proxy.get(peer_id).into_iter().flatten() {
+                    safe.insert(key, peer_id);
+                }
+            }
+        }
+        // Direct TIER1 connections have priority.
+        for key in &accounts_data.keys {
+            if let Some(conn) = tier1.ready_by_account_key.get(&key) {
+                safe.insert(key, &conn.peer_info.id);
+            }
+        }
+
+        // Construct a safe set of connections.
+        let mut safe_set: HashSet<PeerId> = safe.values().map(|v| (*v).clone()).collect();
+        // Add proxies of our node to the safe set.
+        if let Some(vc) = validator_cfg {
+            match &vc.proxies {
+                config::ValidatorProxies::Dynamic(_) => {
+                    safe_set.insert(self.config.node_id());
+                }
+                config::ValidatorProxies::Static(peer_addrs) => {
+                    // TODO(gprusak): here we add peer_id to a safe set, even if
+                    // the conn.peer_addr doesn't match the address from the validator config
+                    // (so we cannot advertise it as our proxy). Consider making it more precise.
+                    safe_set.extend(peer_addrs.iter().map(|pa| pa.peer_id.clone()));
+                }
+            }
+        }
+        // Close all other connections, as they are redundant or are no longer TIER1.
+        for conn in tier1.ready.values() {
+            if !safe_set.contains(&conn.peer_info.id) {
+                conn.stop(None);
+            }
+        }
+        if let Some(vc) = validator_cfg {
+            // Try to establish new TIER1 connections to accounts in random order.
+            let mut handles = vec![];
+            let mut account_keys: Vec<_> = proxies_by_account.keys().copied().collect();
+            account_keys.shuffle(&mut rand::thread_rng());
+            for account_key in account_keys {
+                // tier1_establish_proxies() is responsible for connecting to proxies
+                // of this node. tier1_establish_connections() connects only to proxies
```

Comment:
comment is a little bit confusing, as the current method is called 'tier1_connect'

Reply:
right, fixed